File: go/dto.go
package main

// Mirrors org.acme.dto.AddressDTO. Field order matches the Java record's declaration order, same
// convention used by the Node.js and Rust modules - both differ harmlessly from
// quarkus3-virtual's alphabetical order (its Jackson reflection-free serializers sort keys; see
// nodejs/README.md, "JSON key order differs ... harmlessly"). encoding/json marshals struct
// fields in declaration order, same as serde in the Rust module.
type AddressDto struct {
	Address string `json:"address"`
	City    string `json:"city"`
	Country string `json:"country"`
}

// Mirrors org.acme.dto.StoreDTO.
type StoreDto struct {
	ID       int64      `json:"id"`
	Name     string     `json:"name"`
	Currency string     `json:"currency"`
	Address  AddressDto `json:"address"`
}

// Mirrors org.acme.dto.StoreFruitPriceDTO. Price is never omitted even when zero - it is not
// "empty" for Jackson's NON_EMPTY inclusion, only null/blank/empty-collection values are.
type StoreFruitPriceDto struct {
	Store StoreDto `json:"store"`
	Price float64  `json:"price"`
}

// Mirrors org.acme.dto.FruitDTO.
//
// Description and StorePrices use `omitempty` to replicate
// quarkus.jackson.serialization-inclusion: non-empty from
// quarkus3-virtual/src/main/resources/application.yml - the same rule the Node.js and Rust
// modules implement explicitly in their mapping code. Go's `omitempty` treats an empty string and
// a nil/zero-length slice as empty, which is exactly the behaviour needed here, so no manual
// checks are required the way they are in the other two modules.
type FruitDto struct {
	ID          int64                 `json:"id"`
	Name        string                `json:"name"`
	Description string                `json:"description,omitempty"`
	StorePrices []StoreFruitPriceDto  `json:"storePrices,omitempty"`
}

// Request body for POST /fruits. Mirrors FruitMapper.map(FruitDTO -> Fruit)'s deliberate
// behaviour: only Name and Description are read - id/storePrices sent by a client are ignored,
// same as the Node.js and Rust modules.
//
// Description is a *string (not a plain string) specifically so that an absent field and a
// present-but-empty field are distinguishable: nil means "no description was sent" (persisted as
// SQL NULL, matching Java's Fruit.setDescription(fruitDTO.description()) when the DTO's
// description is null), while a pointer to "" means an explicit empty string was sent. The Rust
// module cannot make this distinction (its DTO field is a non-Option String, so a missing field
// is rejected by serde before validation ever runs - see rust/README.md); this is a case where
// Go achieves closer parity with Java's null-handling than Rust does.
type CreateFruitRequest struct {
	Name        string  `json:"name"`
	Description *string `json:"description,omitempty"`
}

File: rust/src/dto.rs
use serde::{Deserialize, Serialize};

/// Mirrors `org.acme.dto.AddressDTO`. Field order matches the Java record's declaration order,
/// same convention used by the Node.js module (see nodejs/src/dto/address.dto.ts) - both differ
/// harmlessly from quarkus3-virtual's alphabetical order (its Jackson reflection-free serializers
/// sort keys; see nodejs/README.md, "JSON key order differs ... harmlessly").
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AddressDto {
    pub address: String,
    pub city: String,
    pub country: String,
}

/// Mirrors `org.acme.dto.StoreDTO`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StoreDto {
    pub id: i64,
    pub name: String,
    pub currency: String,
    pub address: AddressDto,
}

/// Mirrors `org.acme.dto.StoreFruitPriceDTO`. `price` is never omitted even when zero - it is not
/// "empty" for Jackson's NON_EMPTY inclusion, only null/blank/empty-collection values are.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StoreFruitPriceDto {
    pub store: StoreDto,
    pub price: f64,
}

/// Mirrors `org.acme.dto.FruitDTO`.
///
/// `description` and `store_prices` are skipped when empty to replicate
/// `quarkus.jackson.serialization-inclusion: non-empty` from
/// `quarkus3-virtual/src/main/resources/application.yml` - the same rule the Node.js module
/// implements in its mappers (see nodejs/src/mapping/fruit.mapper.ts).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FruitDto {
    pub id: i64,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub store_prices: Vec<StoreFruitPriceDto>,
}

/// Request body for `POST /fruits`. Only `name` and `description` are accepted - the Java
/// mapper's `FruitMapper.map(FruitDTO -> Fruit)` deliberately maps only these two fields
/// ("the rest of the relationships aren't built out yet"), and both the Node.js and this
/// implementation preserve that on purpose. `id`/`storePrices` sent by a client are ignored.
#[derive(Debug, Clone, Deserialize)]
pub struct CreateFruitRequest {
    pub name: String,
    pub description: Option<String>,
}
